### ISSUE TITLE
Deploy without Elastic IP

### ISSUE THREAD

**Issue by forbode** [boxfuse/boxfuse-issues#48]
Currently I do not see the option to deploy without using an Elastic IP.
Here is my use case:

1) I deploy many micro servers with different services
2) They all communicate via SQS
3) Health checks can be done via Spring Actuator
4) I only need Elastic IP's for the public facing server

I am running out of Elastic IP's (I can ask for more, but I shouldn't need to). 

Is this possible already and I am just missing it?

**Comment by axelfontaine** [comment]
This is not currently possible, but something we have on the agenda. Please ask AWS for an increase of the max number of EIPs for now (they are free when allocated & associated).

Just so we can get a better understanding of your needs: Do you also need these workers behind the SQS queues to be auto-scaled or is it always a single instance? Are they performing long- or short-running jobs? Can they be terminated at any time and restarted, or must they always complete their unit of work in an orderly fashion?

**Comment by forbode** [comment]
>Do you also need these workers behind the SQS queues to be auto-scaled or is it always a single instance? 

Auto-Scaling would be good in the long run, but for now Single Instance. With SQS there is additional understanding I need to have before auto scaling. Though I do know that I would make use of it.

> Are they performing long- or short-running jobs?

Both. One service performs some very short tasks. Another performs a long running (15 Min or less) task. 

>Can they be terminated at any time and restarted, or must they always complete their unit of work in an orderly fashion?

In the future I will be using SWF or some workflow engine to record the state of each task (so that they can be resumed later if a machine goes down. However, it would be ideal to not interrup long running processes in production. The process could be in the middle of performing some very critical work.

**Comment by forbode** [comment]
FYI, I could see a capability where I can use Spring Actuator to produce a metric that says whether or not it's safe to terminate the instance or whether or not it's working on a task. 

However, that is a specific solution. A more general solution might involve examining the CPU to get a sense of current work load. What are your ideas to determine whether or not you can shut down an instance?

**Comment by axelfontaine** [comment]
I like the sound of that. This doesn't have to be Spring Boot specific. It could be a certain path at a certain port that must return HTTP 200 to be protected from termination.

You could then have myhost:1234/stillrunning return either 200 (no killing allowed) or 404 (ok to terminate) for example. Would that work for you?

**Comment by forbode** [comment]
Works for me. Hopefully it works for others. A service like that is very
simple to write if you are monitoring all your Service beans.

**Comment by axelfontaine** [comment]
We have put some more thought into this as we weren't fully happy with requiring a service that listen to queues to have to open a port in the security group for notifications.

Instead this is the model that we are now favoring:
- all workers are always launched in an auto-scaling group (one per version of the app in each environment)
- the autoscaling group has either a fixed capacity (1 in your case) or can auto-scale based on some metric like SQS queue depth or CPU usage
- each "Queue Worker App" is associated with an SQS input queue in each environment, either auto-created by Boxfuse or supplied by you (through boxfuse cfg or the console configure UI)
- the input queue url is injected as an environment variable in each worker app instance upon startup (ex. BOXFUSE_QUEUE_INPUT or something like that) which your app can then use to receive messages from

If you want to delay shutdown of a worker longer by just a few seconds for it to complete work:
- count the number of in-flight worker threads in your app
- add a shutdown hook to the JVM to clean up

If you want to delay shutdown of a worker longer than a few seconds for it to complete work
- define an IAM role to allow auto-scaling to notify your AWS account
- configure Boxfuse with the IAM role name and the maximum delay in seconds (from 0 to 3600 (one hour) )
- before the auto-scaling group gets created, Boxfuse creates a notification queue (a new one for each version of your app in each environment) and also injects the name of this queue as an environment in each instance (ex.: BOXFUSE_QUEUE_SHUTDOWN_NOTIFICATION)
- when a scale-in activity is about to happens, Boxfuse intercepts the auto-scaling activity and forces auto-scaling to place a notification message on that queue
- your app can then choose to react to that notification, stop accepting new work and initiate a shutdown of the instance when ready
- once the maximum delay has elapsed the instance is forcefully terminated

When deploying new versions:
- Boxfuse creates a new auto-scaling group (and a new notification queue if necessary)
- Auto-scaling launches the new workers
- the old auto-scaling group begins to terminate the old workers (and notify them if necessary)
- Boxfuse deletes the old auto-scaling group (and the old notification queue if necessary)


Open questions:
- How to health check the new instances? Also via a pair of SQS queues (are you OK? - yes -> fine, - no -> terminate?) Or simply let the workers healthcheck themselves on startup an commit suicide if the check fails? Boxfuse could then wait until the desired number of workers are running before shutting down the old ones (or kill the new auto-scaling group if the new workers fail to come up)


I know this is a lot to digest, but I would really appreciated your thoughts on this. Does our logic make sense? Did we miss some important aspect? Would this work for you? What are your thoughts on the healthcheck question?

**Comment by forbode** [comment]
It's hard for me to comment on the auto scaling because I haven't begun to research how and when I will need to scale. However, I can try:

Let me first try rephrasing your thoughts... You are hoping that that you can issue a message to a Queue (configurable name) that our apps will intercept and shutdown immediately or with a delay. The only thing our apps will have to do is shutdown or delay, no Queue response message? 

How to health check? Why a queue for healthchecks? What does that by you if you already perform a healthcheck on startup, which you do when I launch a new app. 

I know it's not ideal, but since I am using spring actuator I am a huge fan of incorporating a health metric that indicates whether or not the machine is able to be terminated. However, that's specific and doesn't handle all your other use cases, correct? (I.E. those that are not using actuator or spring).

Bear with me as I understand this. I'll help as much as I can.

**Comment by axelfontaine** [comment]
> You are hoping that that you can issue a message to a Queue (configurable name) that our apps will intercept and shutdown immediately or with a delay. The only thing our apps will have to do is shutdown or delay, no Queue response message?

Correct. This would be a new queue created by Boxfuse that serves only one purpose notifying the app that a forceful shutdown has been scheduled at UTC timestamp X. The app can then choose to react on it by stopping to accept new work, gracefully complete existing work and exit. Exiting the app would lead to automatic instance termination. The app can discover the queue it should listen to for this notification by inspecting the BOXFUSE_QUEUE_SHUTDOWN_NOTIFICATION environment variable.

> How to health check? Why a queue for healthchecks? What does that by you if you already perform a healthcheck on startup, which you do when I launch a new app.

The app should perform a health check on itself on startup and shut down if it fails. This will then cause the instance to auto-terminate

> I know it's not ideal, but since I am using spring actuator I am a huge fan of incorporating a health metric that indicates whether or not the machine is able to be terminated. However, that's specific and doesn't handle all your other use cases, correct? (I.E. those that are not using actuator or spring).

The biggest issue here is that a worker app now needs to be reachable via http(s) as well, which otherwise wouldn't be the case.

--

Did this help to clear things up? What are your thoughts on this?

**Comment by forbode** [comment]
This all sounds acceptable. 
One note. I have processes that might need to last as long as 10 - 20 minutes and cannot be interrupted. So I would need the ability to wait at least the upper bound of 20 minutes before termination.

**Comment by axelfontaine** [comment]
These are our current ideas for this:

Similar to what was described here https://github.com/boxfuse/boxfuse-issues/issues/47#issuecomment-217627840

Except with notification and shutdown delays handles by standard AWS Auto-Scaling Group features.

**Comment by forbode** [comment]
So correct me if I am wrong, this is mostly solved except for the shutdown delays?
